fms/modules/attention.py

```diff
@@ -265,29 +265,56 @@ def __init__(
         assert (
             nheads % world_size == 0
         ), "The number of heads must be divisible by world size"
+        assert (kvheads >= world_size and kvheads % world_size == 0) or (
+            kvheads < world_size and world_size % kvheads == 0
+        ), "the kv heads must be divisible by the world size or the world size must be divisible by kv heads"
         MultiHeadAttention.__init__(
             self,
             emb_dim,
             emb_kq,
             emb_v,
             nheads // world_size,
-            (kvheads // world_size) if kvheads > 1 else kvheads,
+            (kvheads // world_size) if kvheads >= world_size else 1,
             p_dropout,
             use_bias,
             position_encoder,
         )
+        self.pre_tp_nheads = nheads
         self.pre_tp_kvheads = kvheads
         self.setup_tp(rank, world_size)
 
-    def colwise_param_names(self) -> List[str]:
-        colwise_weights = ["query"]
-        if self.pre_tp_kvheads != 1:
-            colwise_weights.append("key")
-            colwise_weights.append("value")
-        return colwise_weights
-
-    def rowwise_param_names(self) -> List[str]:
-        return ["dense"]
+    def load_weights(
+        self,
+        tensor_values: Dict[str, torch.Tensor],
+    ):
+        # 1. Grab the weights from tensor_values
+        query_weight = self._get_sd_weight(tensor_values, ["query", "weight"])
+        key_weight = self._get_sd_weight(tensor_values, ["key", "weight"])
+        value_weight = self._get_sd_weight(tensor_values, ["value", "weight"])
+        dense_weight = self._get_sd_weight(tensor_values, ["dense", "weight"])
+        if self.use_bias:
+            query_bias = self._get_sd_weight(tensor_values, ["query", "bias"])
+            key_bias = self._get_sd_weight(tensor_values, ["key", "bias"])
+            value_bias = self._get_sd_weight(tensor_values, ["value", "bias"])
+            dense_bias = self._get_sd_weight(tensor_values, ["dense", "bias"])
+
+        # 2. Raise exceptions
+        if len(tensor_values) > (8 if self.use_bias else 4):
+            raise AttributeError("Unused weight(s)")
+
+        # 3. Load and shard the weights
+        # The number in max_partition_sizes will signify the largest world size
+        # til we need to duplicate.  For instance if we have nheads=16 and
+        # world_size=32, then first 2 ranks will get first 1/16th of query
+        self.copy_colwise(self.query.weight, query_weight, False, [self.pre_tp_nheads])
```

Comment:
we pass a list for max_partition_sizes (and I guess this replaces what was the value in param_names's dict in a previous revision), but in this PR I don't see anywhere that `len(max_partition_sizes) != 1` (neither for rowwise nor colwise)? where is the use-case for this?

Reply:
next2  PRs in the stack will use the list with size > 1, and mixtral moe TP will too